### PR TITLE
Throw error on markerId update and add test

### PR DESCRIPTION
**Pull request by jzaffiro**
Instead of allowing an update on the markerId and returning undefined when that marker is retrieved using the new id, throw an error when the update is attempted. 

[Ab#3608](https://dev.azure.com/fluidframework/internal/_sprints/taskboard/Team%20Curtis/internal/CY23Q1/Team%20Curtis%20-%202023.03-A?workitem=3608)

### CODE REVIEW

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 1]
```diff
@@ -236,6 +236,29 @@ describe("SharedString", () => {
...
+			assert.equal(simpleMarker.properties?.color, "blue", "Could not annotate marker");
+			// Annotate the marker's ID.
+			const newIdProps = { [reservedMarkerIdKey]: "newIdValue" };
+			try {
+				sharedString.annotateMarker(simpleMarker, newIdProps);
+			} catch (error) {
+				assert(
+					error !== undefined,
+					"Error from attempting to update marker was not thrown",
+				);
```
Comment (connorskees) [inline packages/dds/sequence/src/test/sharedString.spec.ts:258]: More of a nit, but I think `assert.equal` and `assert.notEqual` are preferrable when possible, because the failure message contains the values. When this assert fails, it will only print `false`, which ends up being more painful to debug.

```suggestion
				assert.notEqual(
					error,
					undefined,
					"Error from attempting to update marker was not thrown",
				);
```

I might also check more than just that the error is defined. Comparing against `error.message` or checking `instanceof` would be useful

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 2]
```diff
@@ -236,6 +236,29 @@ describe("SharedString", () => {
...
+			// Annotate the marker.
+			const props = { color: "blue" };
+			const simpleMarker = sharedString.getMarkerFromId("markerId") as Marker;
+			sharedString.annotateMarker(simpleMarker, props);
+			assert.equal(simpleMarker.properties?.color, "blue", "Could not annotate marker");
+			// Annotate the marker's ID.
+			const newIdProps = { [reservedMarkerIdKey]: "newIdValue" };
+			try {
+				sharedString.annotateMarker(simpleMarker, newIdProps);
+			} catch (error) {
```
Comment (connorskees) [inline packages/dds/sequence/src/test/sharedString.spec.ts:254]: It looks like this test can never fail. If `annotateMarker` marker doesn't throw, the `catch` block is never reached. There are a couple ways to solve this; notably you can use `assert.fail` after the `annotateMarker` call to force an exception (note that this would be caught by the `catch`).

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 3]
```diff
@@ -236,6 +236,36 @@ describe("SharedString", () => {
...
+			const newIdProps = { [reservedMarkerIdKey]: "newIdValue" };
+			try {
+				sharedString.annotateMarker(simpleMarker, newIdProps);
+				assert.fail();
+			} catch (error) {
+				assert.notEqual(
+					error,
+					undefined,
+					"Error from attempting to update marker was not thrown",
+				);
```
Comment (alexvy86) [inline packages/dds/sequence/src/test/sharedString.spec.ts:260]: Can this ever happen inside a catch block? I'd expect `error` to always be defined here. However, the "error not thrown" message seems like a good fit for the `assert.fail()` in line 254.

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 4]
```diff
@@ -236,6 +236,36 @@ describe("SharedString", () => {
...
+				assert.equal(
+					(error as Error).message,
+					"Cannot change the markerId of an existing marker",
+					`Incorrect error received - got ${(error as Error).message}`,
+				);
```
Comment (alexvy86) [inline packages/dds/sequence/src/test/sharedString.spec.ts:265]: The intention here is correct, but this way of checking it will break during the next release because our tooling will replace the error message in the `assert()` in `sharedString.ts` with a hex code, so this check will stop seeing the string message. You can use `validateAssertionError` from `@fluidframework/test-runtime-utils` to do it and it'll keep working through that replacement (example [here](https://github.com/microsoft/FluidFramework/blob/a61bd2d371642c3dcadfcdb1ab501faa101e5849/packages/dds/shared-object-base/src/test/sharedObject.spec.ts#L95)).

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 5]
```diff
@@ -249,21 +250,14 @@ describe("SharedString", () => {
...
-				assert.equal(
-					(error as Error).message,
-					"Cannot change the markerId of an existing marker",
-					`Incorrect error received - got ${(error as Error).message}`,
-				);
-			}
+			assert.throws(
+				() => {
+					sharedString.annotateMarker(simpleMarker, newIdProps);
+					assert.fail("Error from attempting to update marker was not thrown");
```
Comment (alexvy86) [inline packages/dds/sequence/src/test/sharedString.spec.ts:256]: Last nit from me 😄 . If we use `assert.throws()` then I think we can skip the `assert.fail()` (`assert.throws()` will already fail the test if the code block doesn't throw, which handles @connorskees 's previous concern), and move this message to the be last parameter to `assert.throws()`, so it prints it as the result of "what I expected to happen (an error that meets the validation function) didn't happen". And if we do, the message probably need an update to `"Error from attempting to update marker was not thrown or was not the expected error"`.

File: packages/dds/sequence/src/sharedString.ts [hunk 6]
```diff
@@ -223,6 +224,7 @@ export class SharedString
 	 * @param combiningOp - Optional. Specifies how to combine values for the property, such as "incr" for increment.
 	 */
 	public annotateMarker(marker: Marker, props: PropertySet, combiningOp?: ICombiningOp) {
+		assert(props.markerId === undefined, "Cannot change the markerId of an existing marker");
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/sharedString.ts:227]: It should also arguably be allowed to set the property to the existing value. I don't have a strong feeling about this though.

Comment (scarlettjlee) [inline packages/dds/sequence/src/sharedString.ts:227]: Why is this check in sharedString? It seems like it should be in mergeTree where the limitation is. Also, not all code paths that could set the markerId are protected here.

Comment (scarlettjlee) [inline packages/dds/sequence/src/sharedString.ts:227]: I think you need to check that the key markerId does not exist at all. Setting the value of the key to undefined will cause PropertyManager to clear the existing property.

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 7]
```diff
@@ -236,6 +237,46 @@ describe("SharedString", () => {
...
+			sharedString.insertText(0, "hello world");
+			// Insert a simple marker.
+			sharedString.insertMarker(6, ReferenceType.Simple, {
+				[reservedMarkerIdKey]: "markerId",
+			});
+			// Annotate the marker's ID.
+			const simpleMarker = sharedString.getMarkerFromId("markerId") as Marker;
+			const newIdProps = { [reservedMarkerIdKey]: "markerId" };
+			sharedString.annotateMarker(simpleMarker, newIdProps);
+			assert.equal(
```
Comment (jzaffiro) [inline packages/dds/sequence/src/test/sharedString.spec.ts:273]: I'm not sure if this is the best condition to check, any suggestions? @scarlettjlee

Comment (scarlettjlee) [reply packages/dds/sequence/src/test/sharedString.spec.ts:273]: Also check the value of the reservedMarkerIdKey property on the marker.

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 8]
```diff
@@ -236,6 +237,46 @@ describe("SharedString", () => {
 			assert.equal(simpleMarker.properties?.color, "blue", "Could not annotate marker");
 		});
 
+		it("fails when the marker is updated", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/sharedString.spec.ts:240]:
```suggestion
		it("fails when the marker id is updated", () => {
```

File: packages/dds/sequence/src/sharedString.ts (hunk 6 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/sharedString.ts]: Add tests for annotating with markerId=null and undefined too.

File: packages/dds/merge-tree/src/opBuilder.ts [hunk 9]
```diff
@@ -31,7 +32,10 @@ export function createAnnotateMarkerOp(
 	if (!id) {
 		return undefined;
 	}
-
+	assert(
+		!props.markerId || props.markerId === id,
+		"Cannot change the markerId of an existing marker",
+	);
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/opBuilder.ts:38]: If you annotate a range (rather than a marker) and the range contains a marker, will it modify the markerId property? I'm not sure it's a valid usage, but it might bypass this check. In which case, it would be better to check in MergeTree.annotateRange's annotateSegment.

Comment (scarlettjlee) [reply packages/dds/merge-tree/src/opBuilder.ts:38]: The marker is a segment. You can call Marker.is on it and get the id.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 10]
```diff
@@ -1911,6 +1911,10 @@ export class MergeTree {
 		let segmentGroup: SegmentGroup | undefined;
 
 		const annotateSegment = (segment: ISegment) => {
+			assert(
+				!(reservedMarkerIdKey in props) || props.markerId === segment.properties?.markerId,
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTree.ts:1915]: Also verify this segment is a marker. It should be allowed to set markerId on a segment that is not a marker and refers to a marker.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 11]
```diff
@@ -1912,7 +1912,9 @@ export class MergeTree {
 
 		const annotateSegment = (segment: ISegment) => {
 			assert(
-				!(reservedMarkerIdKey in props) || props.markerId === segment.properties?.markerId,
+				!(segment.type === "Marker") ||
```
Comment (scarlettjlee) [inline packages/dds/merge-tree/src/mergeTree.ts:1915]: Use Marker.is. That is authoritative rather than duplicating that implementation here. Then you can use segment.getId() below instead of segment.properties?.markerId.

**Review by scarlettjlee — APPROVED** [review]
